File: ios/Classes/YoutubePlayerPlatformView.swift
import Flutter
import UIKit
import WebKit

class YoutubePlayerPlatformView: NSObject, FlutterPlatformView {

    private let webView: WKWebView
    private let methodChannel: FlutterMethodChannel
    private let eventChannel: FlutterEventChannel
    private var eventSink: FlutterEventSink?
    private var isVerticalVideo = false

    init(frame: CGRect, viewId: Int64, messenger: FlutterBinaryMessenger, args: [String: Any]) {
        // -- WKWebView configuration --
        let config = WKWebViewConfiguration()
        config.allowsInlineMediaPlayback = true
        config.mediaTypesRequiringUserActionForPlayback = []

        if #available(iOS 15.4, *) {
            config.preferences.isElementFullscreenEnabled = true
        }

        // JS bridge: creates a YTBridge object that forwards calls to the native handler,
        // so the shared YouTube IFrame HTML (which uses YTBridge.onReady() etc.) works on iOS.
        let userContentController = WKUserContentController()
        let bridgeScript = WKUserScript(
            source: """
            window.YTBridge = {
                onReady: function() {
                    window.webkit.messageHandlers.ytBridge.postMessage({event: 'onReady'});
                },
                onStateChange: function(state) {
                    window.webkit.messageHandlers.ytBridge.postMessage({event: 'onStateChange', state: state});
                },
                onError: function(errorCode) {
                    window.webkit.messageHandlers.ytBridge.postMessage({event: 'onError', errorCode: errorCode});
                },
                onPlaybackQualityChange: function(quality) {
                    window.webkit.messageHandlers.ytBridge.postMessage({event: 'onPlaybackQualityChange', quality: quality});
                }
            };
            document.addEventListener('webkitfullscreenchange', function() {
                var isFS = !!document.webkitFullscreenElement;
                window.webkit.messageHandlers.ytBridge.postMessage({event: 'onFullscreenChange', isFullscreen: isFS});
            });
            document.addEventListener('fullscreenchange', function() {
                var isFS = !!document.fullscreenElement;
                window.webkit.messageHandlers.ytBridge.postMessage({event: 'onFullscreenChange', isFullscreen: isFS});
            });
            """,
            injectionTime: .atDocumentStart,
            forMainFrameOnly: false
        )
        userContentController.addUserScript(bridgeScript)
        config.userContentController = userContentController

        webView = WKWebView(frame: frame, configuration: config)

        methodChannel = FlutterMethodChannel(name: "youtube_player_\(viewId)", binaryMessenger: messenger)
        eventChannel = FlutterEventChannel(name: "youtube_player_events_\(viewId)", binaryMessenger: messenger)

        super.init()

        // Register message handler (weak wrapper to avoid retain cycle)
        userContentController.add(WeakScriptMessageHandler(delegate: self), name: "ytBridge")

        // Setup channels
        methodChannel.setMethodCallHandler { [weak self] call, result in
            self?.handleMethodCall(call: call, result: result)
        }
        eventChannel.setStreamHandler(self)

        // Configure WebView appearance
        webView.backgroundColor = .black
        webView.isOpaque = false
        webView.scrollView.isScrollEnabled = false
        webView.scrollView.bounces = false

        // Parse creation params
        let videoId = args["videoId"] as? String ?? ""
        let autoPlay = args["autoPlay"] as? Bool ?? false
        let showControls = args["showControls"] as? Bool ?? true
        let startSeconds = args["startSeconds"] as? Int ?? 0
        let endSeconds = args["endSeconds"] as? Int ?? 0
        let showFullscreenButton = args["showFullscreenButton"] as? Bool ?? true
        let showRelatedVideos = args["showRelatedVideos"] as? Bool ?? false
        let loop = args["loop"] as? Bool ?? false

        // Generate and load HTML
        let html = Self.generateHTML(
            videoId: videoId,
            autoPlay: autoPlay,
            showControls: showControls,
            startSeconds: startSeconds,
            endSeconds: endSeconds,
            showFullscreenButton: showFullscreenButton,
            showRelatedVideos: showRelatedVideos,
            loop: loop
        )
        // baseURL must match the host used in the IFrame to avoid postMessage origin mismatch
        webView.loadHTMLString(html, baseURL: URL(string: "https://www.youtube-nocookie.com"))

        // Detect if the video is a YouTube Short
        detectIfShort(videoId: videoId)
    }

    func view() -> UIView {
        return webView
    }

    // MARK: - HTML Generation (mirrors KMP YoutubeIFrameGenerator)

    static func generateHTML(
        videoId: String,
        autoPlay: Bool,
        showControls: Bool,
        startSeconds: Int,
        endSeconds: Int,
        showFullscreenButton: Bool,
        showRelatedVideos: Bool,
        loop: Bool
    ) -> String {
        let safeVideoId = videoId.filter { $0.isLetter || $0.isNumber || $0 == "-" || $0 == "_" }
        let autoPlayVal = autoPlay ? 1 : 0
        let controlsVal = showControls ? 1 : 0
        let fsVal = showFullscreenButton ? 1 : 0
        let relVal = showRelatedVideos ? 1 : 0
        let loopVal = loop ? 1 : 0
        let playlist = loop ? "'playlist': '\(safeVideoId)'," : ""
        let start = startSeconds > 0 ? "'start': \(startSeconds)," : ""
        let end = endSeconds > 0 ? "'end': \(endSeconds)," : ""

        return """
        <!DOCTYPE html>
        <html>
        <head>
        <meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=1.0,user-scalable=no">
        <style>*{margin:0;padding:0;overflow:hidden}html,body{width:100%;height:100%;background:#000}#player{width:100%;height:100%}</style>
        </head>
        <body>
        <div id="player"></div>
        <script>
        var tag=document.createElement('script');
        tag.src='https://www.youtube.com/iframe_api';
        document.head.appendChild(tag);
        var player;
        function onYouTubeIframeAPIReady(){
        player=new YT.Player('player',{
        height:'100%',width:'100%',
        videoId:'\(safeVideoId)',
        host:'https://www.youtube-nocookie.com',
        playerVars:{
        'autoplay':\(autoPlayVal),'controls':\(controlsVal),'fs':\(fsVal),
        'enablejsapi':1,'rel':\(relVal),'loop':\(loopVal),
        'playsinline':1,
        \(playlist) \(start) \(end)
        },
        events:{
        'onReady':function(e){try{YTBridge.onReady()}catch(x){}},
        'onStateChange':function(e){try{YTBridge.onStateChange(e.data)}catch(x){}},
        'onError':function(e){try{YTBridge.onError(e.data)}catch(x){}},
        'onPlaybackQualityChange':function(e){try{YTBridge.onPlaybackQualityChange(e.data)}catch(x){}}
        }
        });
        }
        function ytPlay(){if(player)player.playVideo()}
        function ytPause(){if(player)player.pauseVideo()}
        function ytSeekTo(s){if(player)player.seekTo(s,true)}
        function ytLoadVideo(id,s){if(player){player.loadVideoById(id,s||0)}}
        function ytCueVideo(id,s){if(player){player.cueVideoById(id,s||0)}}
        function ytSetQuality(q){if(player)player.setPlaybackQuality(q)}
        function ytSetVolume(v){if(player)player.setVolume(v)}
        function ytMute(){if(player)player.mute()}
        function ytUnMute(){if(player)player.unMute()}
        function ytGetCurrentTime(){return player?player.getCurrentTime():0}
        function ytGetDuration(){return player?player.getDuration():0}
        function ytGetQuality(){return player?player.getPlaybackQuality():'default'}
        function ytGetAvailableQualities(){return player?JSON.stringify(player.getAvailableQualityLevels()):'[]'}
        </script>
        </body>
        </html>
        """
    }

    // MARK: - Shorts Detection

    private func detectIfShort(videoId: String) {
        let safeId = videoId.filter { $0.isLetter || $0.isNumber || $0 == "-" || $0 == "_" }
        guard let url = URL(string: "https://www.youtube.com/shorts/\(safeId)") else { return }

        var request = URLRequest(url: url)
        request.httpMethod = "HEAD"
        request.timeoutInterval = 5
        request.setValue("Mozilla/5.0", forHTTPHeaderField: "User-Agent")

        let delegate = NoRedirectDelegate()
        let session = URLSession(configuration: .ephemeral, delegate: delegate, delegateQueue: nil)

        let task = session.dataTask(with: request) { [weak self] _, response, _ in
            guard let httpResponse = response as? HTTPURLResponse else { return }
            // 200 = it's a Short, 302/303 = redirect to /watch (not a Short)
            let isShort = httpResponse.statusCode == 200

            DispatchQueue.main.async {
                self?.isVerticalVideo = isShort
                self?.eventSink?([
                    "event": "onVideoSizeDetected",
                    "width": isShort ? 9 : 16,
                    "height": isShort ? 16 : 9,
                    "isShort": isShort
                ] as [String: Any])
            }
        }
        task.resume()
    }

    // MARK: - Fullscreen Orientation

    private func setOrientation(landscape: Bool) {
        if #available(iOS 16.0, *) {
            let orientations: UIInterfaceOrientationMask = landscape ? .landscape : .portrait
            guard let windowScene = webView.window?.windowScene
                    ?? UIApplication.shared.connectedScenes.first as? UIWindowScene else { return }
            windowScene.requestGeometryUpdate(.iOS(interfaceOrientations: orientations))
            windowScene.keyWindow?.rootViewController?.setNeedsUpdateOfSupportedInterfaceOrientations()
        } else {
            let orientation: UIInterfaceOrientation = landscape ? .landscapeRight : .portrait
            UIDevice.current.setValue(orientation.rawValue, forKey: "orientation")
        }
    }

    private func restoreOrientation() {
        if #available(iOS 16.0, *) {
            guard let windowScene = webView.window?.windowScene
                    ?? UIApplication.shared.connectedScenes.first as? UIWindowScene else { return }
            windowScene.requestGeometryUpdate(.iOS(interfaceOrientations: .all))
            windowScene.keyWindow?.rootViewController?.setNeedsUpdateOfSupportedInterfaceOrientations()
        } else {
            UIDevice.current.setValue(UIInterfaceOrientation.portrait.rawValue, forKey: "orientation")
        }
    }

    // MARK: - Method Channel Handler

    private func handleMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]

        switch call.method {
        case "play":
            webView.evaluateJavaScript("window.ytPlay && window.ytPlay();")
            result(nil)
        case "pause":
            webView.evaluateJavaScript("window.ytPause && window.ytPause();")
            result(nil)
        case "seekTo":
            let seconds = (args?["seconds"] as? NSNumber)?.doubleValue ?? 0.0
            webView.evaluateJavaScript("window.ytSeekTo && window.ytSeekTo(\(seconds));")
            result(nil)
        case "loadVideo":
            let videoId = args?["videoId"] as? String ?? ""
            let safeId = videoId.filter { $0.isLetter || $0.isNumber || $0 == "-" || $0 == "_" }
            let startSeconds = (args?["startSeconds"] as? NSNumber)?.doubleValue ?? 0.0
            webView.evaluateJavaScript("window.ytLoadVideo && window.ytLoadVideo('\(safeId)', \(startSeconds));")
            detectIfShort(videoId: safeId)
            result(nil)
        case "cueVideo":
            let videoId = args?["videoId"] as? String ?? ""
            let safeId = videoId.filter { $0.isLetter || $0.isNumber || $0 == "-" || $0 == "_" }
            let startSeconds = (args?["startSeconds"] as? NSNumber)?.doubleValue ?? 0.0
            webView.evaluateJavaScript("window.ytCueVideo && window.ytCueVideo('\(safeId)', \(startSeconds));")
            detectIfShort(videoId: safeId)
            result(nil)
        case "setPlaybackQuality":
            let quality = args?["quality"] as? String ?? "default"
            let safeQuality = quality.filter { $0.isLetter || $0.isNumber }
            webView.evaluateJavaScript("window.ytSetQuality && window.ytSetQuality('\(safeQuality)');")
            result(nil)
        case "setVolume":
            let volume = (args?["volume"] as? NSNumber)?.intValue ?? 100
            let clampedVolume = max(0, min(100, volume))
            webView.evaluateJavaScript("window.ytSetVolume && window.ytSetVolume(\(clampedVolume));")
            result(nil)
        case "mute":
            webView.evaluateJavaScript("window.ytMute && window.ytMute();")
            result(nil)
        case "unMute":
            webView.evaluateJavaScript("window.ytUnMute && window.ytUnMute();")
            result(nil)
        case "getCurrentTime":
            webView.evaluateJavaScript("window.ytGetCurrentTime ? window.ytGetCurrentTime() : 0;") { value, _ in
                result(value as? Double ?? 0.0)
            }
        case "getDuration":
            webView.evaluateJavaScript("window.ytGetDuration ? window.ytGetDuration() : 0;") { value, _ in
                result(value as? Double ?? 0.0)
            }
        case "getPlaybackQuality":
            webView.evaluateJavaScript("window.ytGetQuality ? window.ytGetQuality() : 'default';") { value, _ in
                result(value as? String ?? "default")
            }
        case "getAvailableQualityLevels":
            webView.evaluateJavaScript("window.ytGetAvailableQualities ? window.ytGetAvailableQualities() : '[]';") { value, _ in
                result(value as? String ?? "[]")
            }
        case "exitFullscreen":
            webView.evaluateJavaScript(
                "if(document.exitFullscreen){document.exitFullscreen();}else if(document.webkitExitFullscreen){document.webkitExitFullscreen();}"
            )
            restoreOrientation()
            eventSink?(["event": "onFullscreenChange", "isFullscreen": false] as [String: Any])
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Cleanup

    deinit {
        methodChannel.setMethodCallHandler(nil)
        eventChannel.setStreamHandler(nil)
        webView.configuration.userContentController.removeScriptMessageHandler(forName: "ytBridge")
        webView.stopLoading()
    }
}

// MARK: - FlutterStreamHandler

extension YoutubePlayerPlatformView: FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

// MARK: - WKScriptMessageHandler

extension YoutubePlayerPlatformView: WKScriptMessageHandler {
    func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        guard let body = message.body as? [String: Any],
              let event = body["event"] as? String else { return }

        switch event {
        case "onReady":
            eventSink?(["event": "onReady"] as [String: Any])
        case "onStateChange":
            if let state = body["state"] as? Int {
                eventSink?(["event": "onStateChange", "state": state] as [String: Any])
            }
        case "onError":
            if let errorCode = body["errorCode"] as? Int {
                eventSink?(["event": "onError", "errorCode": errorCode] as [String: Any])
            }
        case "onPlaybackQualityChange":
            if let quality = body["quality"] as? String {
                eventSink?(["event": "onPlaybackQualityChange", "quality": quality] as [String: Any])
            }
        case "onFullscreenChange":
            if let isFullscreen = body["isFullscreen"] as? Bool {
                if isFullscreen {
                    setOrientation(landscape: !isVerticalVideo)
                } else {
                    restoreOrientation()
                }
                eventSink?(["event": "onFullscreenChange", "isFullscreen": isFullscreen] as [String: Any])
            }
        default:
            break
        }
    }
}

// MARK: - Weak Script Message Handler (avoids WKWebView retain cycle)

private class WeakScriptMessageHandler: NSObject, WKScriptMessageHandler {
    weak var delegate: WKScriptMessageHandler?

    init(delegate: WKScriptMessageHandler) {
        self.delegate = delegate
        super.init()
    }

    func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        delegate?.userContentController(userContentController, didReceive: message)
    }
}

// MARK: - No-Redirect Session Delegate (for Shorts detection)

private class NoRedirectDelegate: NSObject, URLSessionTaskDelegate {
    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest request: URLRequest,
        completionHandler: @escaping (URLRequest?) -> Void
    ) {
        // Don't follow redirects — a redirect means it's NOT a Short
        completionHandler(nil)
    }
}

File: android/src/main/kotlin/com/openpass/youtube_player_for_flutter/YoutubePlayerPlatformView.kt
package com.openpass.youtube_player_for_flutter

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.pm.ActivityInfo
import android.view.View
import android.webkit.CookieManager
import android.webkit.JavascriptInterface
import android.webkit.WebChromeClient
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.FrameLayout
import com.openpass.shared.PlayerConfig
import com.openpass.shared.YoutubeIFrameGenerator
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView
import java.net.HttpURLConnection
import java.net.URL

@SuppressLint("SetJavaScriptEnabled")
class YoutubePlayerPlatformView(
    context: Context,
    private val viewId: Int,
    messenger: BinaryMessenger,
    params: Map<*, *>,
    private val activity: Activity?
) : PlatformView, MethodChannel.MethodCallHandler {

    private val webView: WebView
    private val methodChannel: MethodChannel
    private val eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null

    private var fullscreenView: View? = null
    private var fullscreenCallback: WebChromeClient.CustomViewCallback? = null
    private var originalSystemUiVisibility: Int = 0
    private var originalOrientation: Int = ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED
    private var isVerticalVideo: Boolean = false

    init {
        // Setup channels
        methodChannel = MethodChannel(messenger, "youtube_player_$viewId")
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(messenger, "youtube_player_events_$viewId")
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })

        // Parse config from creation params
        val videoId = params["videoId"] as? String ?: ""
        val autoPlay = params["autoPlay"] as? Boolean ?: false
        val showControls = params["showControls"] as? Boolean ?: true
        val startSeconds = (params["startSeconds"] as? Number)?.toInt() ?: 0
        val endSeconds = (params["endSeconds"] as? Number)?.toInt() ?: 0
        val showFullscreenButton = params["showFullscreenButton"] as? Boolean ?: true
        val showRelatedVideos = params["showRelatedVideos"] as? Boolean ?: false
        val loop = params["loop"] as? Boolean ?: false

        val config = PlayerConfig(
            videoId = videoId,
            autoPlay = autoPlay,
            showControls = showControls,
            startSeconds = startSeconds,
            endSeconds = endSeconds,
            showFullscreenButton = showFullscreenButton,
            showRelatedVideos = showRelatedVideos,
            loop = loop,
        )

        // Enable cookies
        val cookieManager = CookieManager.getInstance()
        cookieManager.setAcceptCookie(true)

        // Setup WebView
        webView = WebView(context).apply {
            settings.javaScriptEnabled = true
            settings.mediaPlaybackRequiresUserGesture = false
            settings.domStorageEnabled = true
            settings.loadWithOverviewMode = true
            settings.useWideViewPort = true
            settings.mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
            settings.cacheMode = WebSettings.LOAD_DEFAULT

            setBackgroundColor(0xFF000000.toInt())
            isVerticalScrollBarEnabled = false
            isHorizontalScrollBarEnabled = false

            webViewClient = WebViewClient()
            webChromeClient = object : WebChromeClient() {
                override fun onShowCustomView(view: View?, callback: CustomViewCallback?) {
                    if (fullscreenView != null || view == null) {
                        callback?.onCustomViewHidden()
                        return
                    }

                    val act = activity ?: return

                    fullscreenView = view
                    fullscreenCallback = callback

                    originalOrientation = act.requestedOrientation

                    @Suppress("DEPRECATION")
                    originalSystemUiVisibility = act.window.decorView.systemUiVisibility

                    @Suppress("DEPRECATION")
                    act.window.decorView.systemUiVisibility = (
                        View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                            or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                            or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                            or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                            or View.SYSTEM_UI_FLAG_FULLSCREEN
                            or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                    )

                    val decorView = act.window.decorView as FrameLayout
                    decorView.addView(
                        view,
                        FrameLayout.LayoutParams(
                            FrameLayout.LayoutParams.MATCH_PARENT,
                            FrameLayout.LayoutParams.MATCH_PARENT
                        )
                    )

                    act.requestedOrientation = if (isVerticalVideo) {
                        ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
                    } else {
                        ActivityInfo.SCREEN_ORIENTATION_SENSOR_LANDSCAPE
                    }

                    eventSink?.success(mapOf("event" to "onFullscreenChange", "isFullscreen" to true))
                }

                override fun onHideCustomView() {
                    exitFullscreen()
                }
            }

            addJavascriptInterface(YTBridge(), "YTBridge")
        }

        cookieManager.setAcceptThirdPartyCookies(webView, true)

        // Load HTML with IFrame API
        // baseURL must match what YouTube expects for postMessage origin
        val html = YoutubeIFrameGenerator.generate(config)
        webView.loadDataWithBaseURL(
            "https://www.youtube-nocookie.com",
            html,
            "text/html",
            "UTF-8",
            null
        )

        // Detect if the video is a Short from native side
        detectIfShort(videoId)
    }

    override fun getView(): View = webView

    private fun detectIfShort(videoId: String) {
        Thread {
            try {
                val url = URL("https://www.youtube.com/shorts/$videoId")
                val connection = url.openConnection() as HttpURLConnection
                connection.instanceFollowRedirects = false
                connection.requestMethod = "HEAD"
                connection.connectTimeout = 5000
                connection.readTimeout = 5000
                connection.setRequestProperty("User-Agent", "Mozilla/5.0")
                val responseCode = connection.responseCode
                connection.disconnect()

                // 200 = it's a short, 302/303 = redirect to /watch (not a short)
                val isShort = responseCode == 200
                isVerticalVideo = isShort

                webView.post {
                    eventSink?.success(
                        mapOf(
                            "event" to "onVideoSizeDetected",
                            "width" to if (isShort) 9 else 16,
                            "height" to if (isShort) 16 else 9,
                            "isShort" to isShort
                        )
                    )
                }
            } catch (_: Exception) {
                // Default: assume regular video
            }
        }.start()
    }

    private fun exitFullscreen() {
        if (fullscreenView == null) return

        val act = activity ?: return

        val decorView = act.window.decorView as FrameLayout
        decorView.removeView(fullscreenView)

        @Suppress("DEPRECATION")
        act.window.decorView.systemUiVisibility = originalSystemUiVisibility

        fullscreenCallback?.onCustomViewHidden()
        fullscreenView = null
        fullscreenCallback = null

        act.requestedOrientation = originalOrientation

        eventSink?.success(mapOf("event" to "onFullscreenChange", "isFullscreen" to false))
    }

    override fun dispose() {
        exitFullscreen()
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        webView.removeJavascriptInterface("YTBridge")
        webView.stopLoading()
        webView.destroy()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "play" -> {
                webView.evaluateJavascript("window.ytPlay && window.ytPlay();", null)
                result.success(null)
            }
            "pause" -> {
                webView.evaluateJavascript("window.ytPause && window.ytPause();", null)
                result.success(null)
            }
            "seekTo" -> {
                val seconds = call.argument<Number>("seconds")?.toDouble() ?: 0.0
                webView.evaluateJavascript("window.ytSeekTo && window.ytSeekTo($seconds);", null)
                result.success(null)
            }
            "loadVideo" -> {
                val videoId = call.argument<String>("videoId") ?: ""
                val startSeconds = call.argument<Number>("startSeconds")?.toDouble() ?: 0.0
                webView.evaluateJavascript(
                    "window.ytLoadVideo && window.ytLoadVideo('$videoId', $startSeconds);",
                    null
                )
                detectIfShort(videoId)
                result.success(null)
            }
            "cueVideo" -> {
                val videoId = call.argument<String>("videoId") ?: ""
                val startSeconds = call.argument<Number>("startSeconds")?.toDouble() ?: 0.0
                webView.evaluateJavascript(
                    "window.ytCueVideo && window.ytCueVideo('$videoId', $startSeconds);",
                    null
                )
                detectIfShort(videoId)
                result.success(null)
            }
            "setPlaybackQuality" -> {
                val quality = call.argument<String>("quality") ?: "default"
                webView.evaluateJavascript(
                    "window.ytSetQuality && window.ytSetQuality('$quality');",
                    null
                )
                result.success(null)
            }
            "setVolume" -> {
                val volume = call.argument<Number>("volume")?.toInt() ?: 100
                webView.evaluateJavascript(
                    "window.ytSetVolume && window.ytSetVolume($volume);",
                    null
                )
                result.success(null)
            }
            "mute" -> {
                webView.evaluateJavascript("window.ytMute && window.ytMute();", null)
                result.success(null)
            }
            "unMute" -> {
                webView.evaluateJavascript("window.ytUnMute && window.ytUnMute();", null)
                result.success(null)
            }
            "getCurrentTime" -> {
                webView.evaluateJavascript(
                    "window.ytGetCurrentTime ? window.ytGetCurrentTime() : 0;"
                ) { value ->
                    val time = value?.toDoubleOrNull() ?: 0.0
                    result.success(time)
                }
            }
            "getDuration" -> {
                webView.evaluateJavascript(
                    "window.ytGetDuration ? window.ytGetDuration() : 0;"
                ) { value ->
                    val duration = value?.toDoubleOrNull() ?: 0.0
                    result.success(duration)
                }
            }
            "getPlaybackQuality" -> {
                webView.evaluateJavascript(
                    "window.ytGetQuality ? window.ytGetQuality() : 'default';"
                ) { value ->
                    val quality = value?.removeSurrounding("\"") ?: "default"
                    result.success(quality)
                }
            }
            "getAvailableQualityLevels" -> {
                webView.evaluateJavascript(
                    "window.ytGetAvailableQualities ? window.ytGetAvailableQualities() : '[]';"
                ) { value ->
                    val cleaned = value?.removeSurrounding("\"")?.replace("\\\"", "\"") ?: "[]"
                    result.success(cleaned)
                }
            }
            "exitFullscreen" -> {
                exitFullscreen()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private inner class YTBridge {
        @JavascriptInterface
        fun onReady() {
            webView.post {
                eventSink?.success(mapOf("event" to "onReady"))
            }
        }

        @JavascriptInterface
        fun onStateChange(state: Int) {
            webView.post {
                eventSink?.success(mapOf("event" to "onStateChange", "state" to state))
            }
        }

        @JavascriptInterface
        fun onError(errorCode: Int) {
            webView.post {
                eventSink?.success(mapOf("event" to "onError", "errorCode" to errorCode))
            }
        }

        @JavascriptInterface
        fun onPlaybackQualityChange(quality: String) {
            webView.post {
                eventSink?.success(
                    mapOf("event" to "onPlaybackQualityChange", "quality" to quality)
                )
            }
        }
    }
}
